### PR TITLE
feat(linter): implement `unicorn/consistent-existence-index-check` lint rule 

### PR DESCRIPTION
**Pull request by ryan-m-walker**
Hey all, love the project and looking to contribute. Hope this helps!

[unicorn/consistent-existence-index-check.](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/consistent-existence-index-check.md) for #684

### CODE REVIEW

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+    ConsistentExistenceIndexCheck,
+    restriction,
+    fix,
+);
+
+const METHOD_NAMES: [&str; 4] = ["indexOf", "lastIndexOf", "findIndex", "findLastIndex"];
+
+impl Rule for ConsistentExistenceIndexCheck {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::BinaryExpression(binary_expression) = node.kind() {
```
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:63]: prefer `let AstKind::BinaryExpression(binary_expression) = node.kind() else {return; };` please, it helps avoid overly nested blocks

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+            let left = binary_expression.left.without_parentheses();
+            let right = binary_expression.right.without_parentheses();
```
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:65]: prefer using `inner_expression()` as it also ignores stuff like type casts, non null expressions ect

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+            let operator = binary_expression.operator;
+
+            let Expression::Identifier(identifier) = left else {
+                return;
+            };
```
Comment (camc314) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:70]: does this rule not care if the ident is on the right?

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs (hunk 2 above)
Comment (ryan-m-walker) [reply crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:65]: good to know!

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs (hunk 3 above)
Comment (ryan-m-walker) [reply crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:70]: That seems to be the case with the original implementation.

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/consistent-existence-index-check.js#L82-L84

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+            let replacement = get_replacement(right, &operator);
+
+            let Some(replacement) = &replacement else {
+                return;
+            };
+
+            let existence_or_non_existence =
+                if replacement.replacement_value == "-1" { "non-existence" } else { "existence" };
+
+            let label = format!(
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:117]: Let's move this error message logic into the diagnostic. Putting all the diagnostic messages together at the top makes it easier to see what this rule emits.

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+    /// const index = foo.indexOf('bar');
+    /// if (index === -1) {}
+    /// ```
+    ///
+    /// ``` javascript
+    /// const index = foo.indexOf('bar');
+    /// if (index !== -1) {}
+    /// ```
+    ConsistentExistenceIndexCheck,
+    correctness,
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:55]:
```suggestion
    style,
```

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+
+            let operator_matches = operator_source.match_indices(replacement.original_operator);
+            let mut operator_replacement_text = operator_source.to_string();
+
+            for (index, text) in operator_matches {
+                let comments = ctx.semantic().comments_range(operator_start..operator_end);
+
+                let start = operator_start + index as u32;
+                let end = start + text.len() as u32;
+                let span = Span::new(start, end);
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:139]: You can used Span::sized here to start a span at a point for some length

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+                        existenceOrNonExistence = existence_or_non_existence,
+                    );
+
+            let operator_start = binary_expression.left.span().end;
+            let operator_end = binary_expression.right.span().start;
+            let operator_span = Span::new(operator_start, operator_end);
+            let operator_source = ctx.source_range(operator_span);
+
+            let operator_matches = operator_source.match_indices(replacement.original_operator);
+            let mut operator_replacement_text = operator_source.to_string();
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:132]: Let's move all of the fix generation inside of the fixer closure. That way we don't need to run it unless fixes are enabled.

File: crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+        			/* comment 5 */
+        			- /* comment 6 */ (( /* comment 7 */ 1 /* comment 8 */ ))
+        			/* comment 9 */
+        		))
+        	));
+        ",
+        r"const index = _.indexOf([1, 2, 1, 2], 2); index < 0;",
+    ];
+
+    let fix = vec![
```
Comment (camchenry) [inline crates/oxc_linter/src/rules/unicorn/consistent_existence_index_check.rs:375]: Can we add all of the fail cases here as well, to make sure they have the correct output?

**Review by camchenry — COMMENTED** [review]
This looks great! A few small comments but then I think this will be ready

**Comment by ryan-m-walker** [comment]
Thanks for the feedback! think i got all yalls suggestions. I also realized I had some clippy issues so fixed those as well

**Review by camchenry — APPROVED** [review]

**Comment by camchenry** [comment]
### Merge activity

* **Nov 15, 11:59 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

**Pull request merged**